### ISSUE TITLE
Concurrency Error in org.bouncycastle.math.ec.ECAlgorithms.implShamirsTrickWNaf

### ISSUE THREAD

**Issue by gerlion** [bcgit/bc-java#366]
When using the SECP curves under heavy concurrent load multiplication will occasionally fail.  

Caused by: java.lang.ArrayIndexOutOfBoundsException: 1
	at org.bouncycastle.math.ec.ECAlgorithms.implShamirsTrickWNaf(Unknown Source)
	at org.bouncycastle.math.ec.ECAlgorithms.implShamirsTrickWNaf(Unknown Source)
	at org.bouncycastle.math.ec.GLVMultiplier.multiplyPositive(Unknown Source)
	at org.bouncycastle.math.ec.AbstractECMultiplier.multiply(Unknown Source)

The following code, and attached source files, are a fairly minimal example.  Note, that while it fails often it is randomised. 

I am using a MacBook Pro (MacBookPro14,3) running macOS 10.13.4, java version "1.8.0_131" and bcprov-jdk15on-159.jar.

```java
import static java.util.stream.Collectors.toList;

import java.math.BigInteger;
import java.util.List;
import java.util.stream.IntStream;

import org.bouncycastle.jce.ECNamedCurveTable;
import org.bouncycastle.jce.spec.ECNamedCurveParameterSpec;
import org.bouncycastle.math.ec.ECPoint;
import org.bouncycastle.math.ec.custom.sec.SecP256K1Curve;

public class ConcurrencyError
{
	public static final ECNamedCurveParameterSpec group = ECNamedCurveTable.getParameterSpec("secp256k1");		
	public static final SecP256K1Curve curve = new SecP256K1Curve();
	public static final ECPoint g = group.getG();
	public static final BigInteger q = curve.getOrder();
	public static final ECPoint aPoint = g.multiply(BigInteger.valueOf(3277561));

	public static BigInteger order() {	//Order of the group
		return curve.getOrder();
	}
	
	public static void foo() {
		final int N = 50000;
		final int M = 100;
		List<ECPoint> T = IntStream.range(0,N).mapToObj(__ -> aPoint).collect(toList());
		for (int i=0; i<M; ++i) {
			System.out.println("i = " + i);
			T = compute(T);
		}
	}
	
	private static List<ECPoint> compute(List<ECPoint> T) {
		final int N = T.size();
		final List<ECPoint> S = IntStream.range(0, N)
				.mapToObj(i -> T.get(i).add(T.get((i+1)%N)))
				.collect(toList());
		return IntStream.range(0, N).parallel()
				.mapToObj(i -> S.get(0).multiply(exponent(i)))
				.collect(toList());
	}

	public static BigInteger exponent(int i) {
		if (i%2 == 0) {
			return BigInteger.valueOf(7);
		} else {
			return  BigInteger.valueOf(3277561);
		}
	}


	public static void main(String[] args) { 
		foo();
	}
}
```

[BCECError.zip](https://github.com/bcgit/bc-java/files/2098291/BCECError.zip)

**Comment by peterdettman** [comment]
Thanks, I can reproduce the exception here and am investigating. I'm pretty sure it's a race condition in WnafUtil.precompute or a related method, triggered by multiple threads launching a multiply on the same ECPoint simultaneously.

e.g. adding `S.get(0).multiply(exponent(0))` before the `parallel()` task appears to make things work (but of course... random).

**Comment by bcgit** [comment]
There's now a version in https://www.bouncycastle.org/betas for testing as well 160 b14 or later.

**Comment by bcgit** [comment]
Released in 1.60
